File: spark_streaming_demo/src/main/python/word_count_demo.py
#!/usr/bin/env python
# coding=utf-8
from pyspark import SparkContext, SparkConf
'''
./bin/spark-submit --master spark://master:7077 --executor-memory 1G --deploy-mode cluster --total-executor-cores 3 word_count_demo.py
'''
if __name__ == '__main__':
    conf = SparkConf().setAppName('logfile_character_count')
    sc = SparkContext(conf=conf)
    texts = sc.textFile('hdfs://index:8020/wordcount')
    texts.foreach(lambda line:print line)
    counts = texts.flatMap(lambda line:line.split(" ")).map(lambda word:(word, 1)).reduceByKey(lambda a, b:a + b)
    # 转为driver结果
    count_result = counts.collect()
    for result in count_result:
        print result

    print "save to hdfs."
    counts.saveAsTextFile('hdfs://index:8020/wordcount_output')

File: spark_demo/src/main/scala/net/zengzhiying/WordCountDemo.scala
package net.zengzhiying

import org.apache.spark.sql.SparkSession

object WordCountDemo {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession
                .builder()
                .appName("word_count_demo")
//                .master("local")
                .getOrCreate()
    val texts = spark.read.textFile("hdfs://index:8020/wordcount")
    val counts = texts.rdd.flatMap(line => lineSplit(line)).map((_, 1)).reduceByKey(add(_, _))
    counts.collect().foreach(println)
    counts.saveAsTextFile("hdfs://index:8020/wordcount_output")
    spark.close()
  }
  
  def lineSplit(line: String): Array[String]= {
    return line.split(" ")
  }
  
  def add(a: Int, b: Int): Int = {
    return a + b
  }
}